### PR TITLE
Feature/89 add oracle support

### LINKED ISSUE #89: Oracle support
Add support for Oracle

### PR DESCRIPTION
**Pull request by pj-spoelders**
Closes #89

### CODE REVIEW

File: src/main/java/com/exasol/dbbuilder/dialects/oracle/OracleSchema.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    public Table createTable(final String name, final List<String> columnNames, final List<String> columnTypes) {
+        if (columnNames.size() == columnTypes.size()) {
+            //Create a local table builder + enter info
+            final OracleTable.Builder builder = OracleTable.builder(getWriter(), this, getIdentifier(name));
+            passColumnsToTableBuilder(columnNames, columnTypes, builder);
+            //Build a table with the builder
+            final OracleTable table = builder.build();
+            //add the table to the schema's tables list
+            this.tables.add(table);
+            //return the new table object (reference)
```
Comment (kaklakariada) [inline src/main/java/com/exasol/dbbuilder/dialects/oracle/OracleSchema.java:55]: Remove comments, they just duplicate the code

File: src/test/java/com/exasol/dbbuilder/dialects/AbstractDatabaseObjectCreationAndDeletionIT.java [hunk 2]
```diff
@@ -23,7 +23,7 @@
     protected Connection adminConnection;
 
     @BeforeEach
-    void beforeEach() throws SQLException {
+    protected void beforeEach() throws SQLException {
```
Comment (kaklakariada) [inline src/test/java/com/exasol/dbbuilder/dialects/AbstractDatabaseObjectCreationAndDeletionIT.java:26]: This should not be necessary, better use default visibility if possible

File: src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    //Code smell but also how it's done in the class it's inheriting from.
+    // Finding a more novel approach isn't the goal here.
+    // The only purpose of this class is delivering a DBA user connection
+    // (which isn't done by default like for the postgresql or mysql containers).
+    @SuppressWarnings("java:S2925")
+    public Connection createConnectionDBA(String queryString) throws SQLException, JdbcDatabaseContainer.NoDriverFoundException {
+        Properties info = new Properties();
+        info.put("user", "SYSTEM");
+        //info.put("user", this.getUsername());
```
Comment (kaklakariada) [inline src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java:30]:
```suggestion
```

File: src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        String url = this.constructUrlForConnection(queryString);
+        Driver jdbcDriverInstance = this.getJdbcDriverInstance();
+        SQLException lastException = null;
+
+        try {
+            long start = System.currentTimeMillis();
+
+            while(System.currentTimeMillis() < start + (long)(1000 * this.connectTimeoutSeconds) && this.isRunning()) {
+                try {
+                    //this.logger().debug("Trying to create JDBC connection using {} to {} with properties: {}", new Object[]{this.driver.getClass().getName(), url, info});
```
Comment (kaklakariada) [inline src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java:41]: Delete comment

File: src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                    return jdbcDriverInstance.connect(url, info);
+                } catch (SQLException var9) {
+                    lastException = var9;
```
Comment (kaklakariada) [inline src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java:44]: Rename `var9` to `exception`

File: src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleDatabaseObjectCreationAndDeletionIT.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.SQLException;
+import java.util.Properties;
+
+class OracleDatabaseObjectCreationAndDeletionIT extends AbstractDatabaseObjectCreationAndDeletionIT {
+    private static final String ORACLE_DOCKER_IMAGE_REFERENCE = "gvenzl/oracle-xe:21.3.0";
+    private static final String QUOTES = "\"\"";
+    //private static final String QUOTES = "` `` ' '' \" \"\"";
```
Comment (kaklakariada) [inline src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleDatabaseObjectCreationAndDeletionIT.java:25]: Delete comment

File: .classpath [hunk 7]
```diff
@@ -29,5 +29,33 @@
 			<attribute name="test" value="true"/>
 		</attributes>
 	</classpathentry>
+	<classpathentry kind="src" path="target/generated-sources/annotations">
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply .classpath:32]: ok

File: .project [hunk 8]
```diff
@@ -20,4 +20,15 @@
 		<nature>org.eclipse.jdt.core.javanature</nature>
 		<nature>org.eclipse.m2e.core.maven2Nature</nature>
 	</natures>
+	<filteredResources>
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply .project:23]: ok

File: src/main/java/com/exasol/dbbuilder/dialects/Table.java [hunk 9]
```diff
@@ -104,7 +109,7 @@ public Table bulkInsert(final Stream<List<Object>> rows) {
      */
     public static class Builder {
         private final DatabaseObjectWriter writer;
-        private final Identifier name;
+        private final Identifier tableName;
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply src/main/java/com/exasol/dbbuilder/dialects/Table.java:112]: 😅

File: src/test/java/com/exasol/dbbuilder/dialects/AbstractDatabaseObjectCreationAndDeletionIT.java (hunk 2 above)
Comment (pj-spoelders) [reply src/test/java/com/exasol/dbbuilder/dialects/AbstractDatabaseObjectCreationAndDeletionIT.java:26]: fixed

File: src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java (hunk 4 above)
Comment (pj-spoelders) [reply src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java:40]: done

File: src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleDatabaseObjectCreationAndDeletionIT.java (hunk 6 above)
Comment (pj-spoelders) [reply src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleDatabaseObjectCreationAndDeletionIT.java:25]: done

File: src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java (hunk 5 above)
Comment (pj-spoelders) [reply src/test/java/com/exasol/dbbuilder/dialects/oracle/OracleContainerDBA.java:43]: done

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractSchema.java [hunk 10]
```diff
@@ -60,14 +60,14 @@ public Table createTable(final String name, final String column1Name, final Stri
...
-            for (final String columnName : columnNames) {
-                builder.column(columnName, columnTypes.get(index));
-                ++index;
-            }
+            passColumnsToTableBuilder(columnNames, columnTypes, builder);
+            //Build a table with the builder
             final Table table = builder.build();
+            //add the table to the schema's tables list
             this.tables.add(table);
+            //return the new table object (reference)
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply src/main/java/com/exasol/dbbuilder/dialects/AbstractSchema.java]: Removed comments

File: src/main/java/com/exasol/dbbuilder/dialects/oracle/OracleImmediateDatabaseObjectWriter.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public void write(final User user, final GlobalPrivilege... privileges) {
+        throw new UnsupportedOperationException(ExaError.messageBuilder("E-TDBJ-29").message("Creating users with privileges is not implemented in this version of the test-db-builder.").toString());
+    }
+
+    //a schema is linked to a user in oracle, the schema has the same name as the user
+    @Override
+    public void drop(final Schema schema) {
+        writeToObject(schema,"DROP USER " + schema.getName() +" CASCADE");
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply src/main/java/com/exasol/dbbuilder/dialects/oracle/OracleImmediateDatabaseObjectWriter.java]: Done

**Review by tomuben — DISMISSED** [review]

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**